README.md
# fdstat
**fdstat** reports the number of files descriptors allocated by each process.

The information is collected from the `/proc` filesystem, therefore an
open file may be a regular file, a directory, a block special file, a
character special file, an executing text reference, a library, a
stream or a network file (Internet socket, NFS file or UNIX domain
socket).

Each report provides information gathered at the time of execution.


## Installation

### Debian package
* Download the Debian package from the latest release 
[here](https://github.com/marcoalmeida/fdstat/releases)
* Run 
  
  `sudo dpkg -i fdstat_<version>_amd64.deb`
 
### From source
* Download the tarball with the source code from the latest
release [here](https://github.com/marcoalmeida/fdstat/releases)
* Extract the tarball
* Change to the directory containing the source code:
  
  `cd fdstat-<version>`
  
* Run

  `make && sudo make install`


## Usage

Usage: `fdstat [OPTION...] [INTERVAL [COUNT]]`

Options:
```
  -c, --cmdline              Output the command line instead of the PID

  -j, --json                 Produce JSON output

  -s, --summary              Display only totals (allocated and maximum)

  -v, --verbose              Produce verbose output

  -?, --help                 Give this help list

      --usage                Give a short usage message

  -V, --version              Print program version
```

The `INTERVAL` parameter specifies the amount of time in seconds between each  report.
The  `COUNT` parameter can be specified in conjunction with the `INTERVAL` parameter. If
the `COUNT` parameter is specified, the  value  of `COUNT`  determines  the  number  of
reports  generated at `INTERVAL` seconds apart. If the `INTERVAL` parameter is specified
without the `COUNT` parameter, the **fdstat** command generates reports continuously.

## Examples
* List of number of currently open files and corresponding PID

  `fdstat`

* Continuously list of number of currently open files and corresponding PID,
generating a new report every 3 seconds

  `fdstat 3`

* Print three reports, five seconds apart, on open files and corresponding 
command line 

  `fdstat 5 3 -c`

* Show the full command line, sort the results in descending order (process 
with highest number of open files last) 

  `fdstat -c | sort -k1 -n`

Report bugs to <marcoafalmeida@gmail.com>.


debian/changelog
fdstat (0.2.0-1) unstable; urgency=medium

  * Change output format
  * Install/uninstall targets on the Makefile
  * Issue template
  * Pull request template

 -- Marco Almeida <marcoafalmeida@gmail.com>  Wed, 29 Nov 2017 00:09:14 -0800

fdstat (0.1.1-1) unstable; urgency=medium

  * Fix summary mode
  * Untabify

 -- Marco Almeida <marcoafalmeida@gmail.com>  Sat, 25 Nov 2017 03:00:43 -0800

fdstat (0.1.0-1) unstable; urgency=medium

  * Initial release

 -- Marco Almeida <marcoafalmeida@gmail.com>  Sat, 25 Nov 2017 00:07:35 -0800


main.c
/* Copyright (C) 2017 Marco Almeida <marcoafalmeida@gmail.com> */

/* This file is part of fdstat. */

/* fdstat is free software; you can redistribute it and/or modify */
/* it under the terms of the GNU General Public License as published by */
/* the Free Software Foundation; either version 2 of the License, or */
/* (at your option) any later version. */

/* fdstat is distributed in the hope that it will be useful, */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
/* GNU General Public License for more details. */

/* You should have received a copy of the GNU General Public License */
/* along with fdstat; if not, write to the Free Software */
/* Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */


#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <dirent.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <argp.h>

#include "arg_parser.h"
#include "util.h"


#define _POSIX_C_SOURCE >= 200809L


int count_fds(const char *pid)
{
    char path[MAX_PATH];
    int count = 0;
    DIR *dp;
    struct dirent *entry;

    if (sprintf(path, "/proc/%s/fd", pid) < 0)
        suicide("creating /proc/<pid>/fd path");

    dp = opendir(path);
    /* some processes are very short-lived; just ignore it in case it
     * happens to disappear */
    if (dp) {
        while((entry=readdir(dp))) {
            if (strcmp(entry->d_name, ".") && strcmp(entry->d_name, ".."))
                count++;
        }
    }
    closedir(dp);

    return count;
}

void count_proc_fds(int *procs)
{
    DIR *dp;
    struct dirent *entry;
    struct stat statbuf;

    dp = opendir("/proc");
    if (!dp)
        suicide("reading /proc");

    while((entry=readdir(dp))) {
        stat(entry->d_name, &statbuf);
        if(S_ISDIR(statbuf.st_mode)) {
            if (is_integer(entry->d_name)) {
                procs[atoi(entry->d_name)] = count_fds(entry->d_name);
            }
        }
    }

    closedir(dp);
}

void print(int *proc_fds, int n_procs, int json, int cmdline)
{
    int pid;
    int i_printed = 0;
    char cmdline_buf[MAX_CMDLINE];
    char *begin = json ? "{" : "#FDs\tPID\n";
    char *end = json ? "}\n" : "\n";
    char *pid_fmt = json ? "\"%d\": %d" : "%d\t%d\n";
    char *cmdline_fmt = json ? "\"%s\": %d" : "%d\t%s\n";

    printf(begin, "");
    for (pid = 1; pid < n_procs; pid++) {
        /* ignore processes with no open files */
        if (proc_fds[pid]) {
            /* avoid a trailing comma -- we don't know when the last
             * line will be printed, so keep track of the first one
             * and prepend commas instead of appending */
            if (json && i_printed)
                printf(", ");
            if (cmdline) {
                if (proc_cmdline(pid, cmdline_buf, MAX_CMDLINE)) {
                    /* if JSON, we want the key to be the cmd line */
                    if (json)
                        printf(cmdline_fmt, cmdline_buf, proc_fds[pid]);
                    else
                        printf(cmdline_fmt, proc_fds[pid], cmdline_buf);
                }
                else
                    fprintf(stderr, "Failed to get cmdline for pid %d\n", pid);
            }
            else {
                /* if JSON, we want the key to be the PID */
                if (json)
                    printf(pid_fmt, pid, proc_fds[pid]);
                else
                    printf(pid_fmt, proc_fds[pid], pid);
            }
            i_printed = 1;
        }
    }
    printf(end, "");
}

void print_summary(int allocated, int total, int json) {
    if (json)
        printf("{\"allocated\": %d, \"total\": %d}\n", allocated, total);
    else {
        printf("Allocated\tMaximum available\n");
        printf("%d\t\t%d\n", allocated, total);
    }
}

int main(int argc, char *argv[])
{
    /* array to store the number of open file descriptors per process
     * and total number of running procs */
    int *proc_fds, n_procs=max_pid()+1;
    /* struct for storing argument's values; also initialize/set
     * defaults (conveniently all zero) */
    struct arguments arguments = { 0 };

    /* parse arguments */
    parse_args(argc, argv, &arguments);

    /* allocate the array to store the number of open file descriptors
     * per process */
    if (! (proc_fds=(int *)calloc(n_procs, sizeof(int))))
        suicide("allocating process table");

    while (1) {
        if (arguments.summary) {
            print_summary(allocated_fd(), max_fd(), arguments.json);
            break;
        }
        else {
            /* reset the process array */
            memset(proc_fds, 0, n_procs*sizeof(int));
            /* collect/count the number of file descriptors per process */
            count_proc_fds(proc_fds);
            /* print out the results */
            print(proc_fds, n_procs, arguments.json, arguments.cmdline);
        }

        /* if a COUNT was set, make sure we stop after those many */
        if (arguments.n_args == 2) {
            if (arguments.args[1]-- == 1)
                break;
        }
        /* if an INTERVAL was set, sleep for that period of time */
        if (arguments.n_args > 0)
            sleep(arguments.args[0]);
        /* if neither COUNT nor INTERVAL are set, run just once */
        if (arguments.n_args == 0)
            break;
    }

    /* release memory allocated for */
    free(proc_fds);

    return 0;
}


util.c
/* Copyright (C) 2017 Marco Almeida <marcoafalmeida@gmail.com> */

/* This file is part of fdstat. */

/* fdstat is free software; you can redistribute it and/or modify */
/* it under the terms of the GNU General Public License as published by */
/* the Free Software Foundation; either version 2 of the License, or */
/* (at your option) any later version. */

/* fdstat is distributed in the hope that it will be useful, */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
/* GNU General Public License for more details. */

/* You should have received a copy of the GNU General Public License */
/* along with fdstat; if not, write to the Free Software */
/* Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */


#include <stdio.h>
#include <string.h>
#include <strings.h>
#include <stdlib.h>
#include <ctype.h>
#include <errno.h>

#include "util.h"


void suicide(const char *action)
{
    fprintf(stderr, "Unrecoverable error %s: %s\n", action, strerror(errno));
    exit(EXIT_FAILURE);
}

int is_integer(const char *str)
{
    while (*str)
        if (! isdigit(*str++))
            return 0;

    return 1;
}

int sys_value(const char *path)
{
    char buf[MAX_BUF_SIZE];
    int value = -1;
    FILE *fp;

    if (! (fp=fopen(path, "r")))
        suicide(path);
    if ((fread(buf, 1, MAX_BUF_SIZE-1, fp) > 0) && (! ferror(fp)))
        value = atoi(buf);
    fclose(fp);

    return value;
}

__inline__ int max_fd()
{
    return sys_value("/proc/sys/fs/file-max");
}

__inline__ int max_pid()
{
    return sys_value("/proc/sys/kernel/pid_max");
}

int allocated_fd()
{
    char buf[MAX_BUF_SIZE];
    char *delim;
    FILE *fp;

    if (! (fp=fopen("/proc/sys/fs/file-nr", "r")))
        suicide("opening /proc/sys/fs/file-nr");
    if (fread(buf, 1, MAX_BUF_SIZE, fp) && (!ferror(fp))) {
        /* this file contains 3 integers: allocated, unused, maxiumum;
         * we want the first */
        if ((delim=index(buf, '\t'))) {
            *delim = 0;
            return atoi(buf);
        }
    }
    fclose(fp);

    return -1;
}

int proc_cmdline(int pid, char *buf, int size)
{
    int success = 1;
    char path[MAX_PATH];
    FILE *fp;

    if (sprintf(path, "/proc/%d/cmdline", pid) < 0)
        suicide("creating /proc/<pid>/cmdline path");

    /* some processes are very short-lived; just ignore it in case it
     * happens to disappear */
    if ((fp=fopen(path, "r"))) {
        if (!fread(buf, 1, size, fp) || ferror(fp))
            success = 0;
        fclose(fp);
    }

    return success;
}


util.h
/* Copyright (C) 2017 Marco Almeida <marcoafalmeida@gmail.com> */

/* This file is part of fdstat. */

/* fdstat is free software; you can redistribute it and/or modify */
/* it under the terms of the GNU General Public License as published by */
/* the Free Software Foundation; either version 2 of the License, or */
/* (at your option) any later version. */

/* fdstat is distributed in the hope that it will be useful, */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
/* GNU General Public License for more details. */

/* You should have received a copy of the GNU General Public License */
/* along with fdstat; if not, write to the Free Software */
/* Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */


#ifndef UTIL_H_
#define UTIL_H_

#define MAX_PATH 320
#define MAX_CMDLINE 8072 /* arbitrary but large enough (?) */
#define MAX_BUF_SIZE 32


/* print out the string that describes errno and exit */
void suicide(const char *action);
/* return 1 if str points to a string of digits, 0 otherwise */
int is_integer(const char *str);
/* return the maximum number of file handles for the entire system */
int max_fd();
/* return the number of total allocated file handles or -1 on error */
int allocated_fd();
/* return the maximum number of processes */
int max_pid();
/* read the cmdline of pid into buf (of size size) */
int proc_cmdline(int pid, char *buf, int size);


#endif /* UTIL_H_ */


arg_parser.h
/* Copyright (C) 2017 Marco Almeida <marcoafalmeida@gmail.com> */

/* This file is part of fdstat. */

/* fdstat is free software; you can redistribute it and/or modify */
/* it under the terms of the GNU General Public License as published by */
/* the Free Software Foundation; either version 2 of the License, or */
/* (at your option) any later version. */

/* fdstat is distributed in the hope that it will be useful, */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
/* GNU General Public License for more details. */

/* You should have received a copy of the GNU General Public License */
/* along with fdstat; if not, write to the Free Software */
/* Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */


#ifndef ARG_PARSER_H_
#define ARG_PARSER_H_


/* Used by main to communicate with parse_opt. */
struct arguments {
    /* number of arguments we received: interval and/or count */
    int n_args;
    int args[2]; /* interval and count */
    int summary, verbose, json, cmdline;
};


void parse_args(int argc, char *argv[], struct arguments *arguments);


#endif /* ARG_PARSER_H_ */


arg_parser.c
/* Copyright (C) 2017 Marco Almeida <marcoafalmeida@gmail.com> */

/* This file is part of fdstat. */

/* fdstat is free software; you can redistribute it and/or modify */
/* it under the terms of the GNU General Public License as published by */
/* the Free Software Foundation; either version 2 of the License, or */
/* (at your option) any later version. */

/* fdstat is distributed in the hope that it will be useful, */
/* but WITHOUT ANY WARRANTY; without even the implied warranty of */
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
/* GNU General Public License for more details. */

/* You should have received a copy of the GNU General Public License */
/* along with fdstat; if not, write to the Free Software */
/* Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */


#include <stdlib.h>
#include <argp.h>

#include "arg_parser.h"


/* put the given documentation string and bug address in the ‘--help’
 * output, as per GNU standards */
const char *argp_program_version = "fdstat 0.2";
const char *argp_program_bug_address = "<marcoafalmeida@gmail.com>";


/* Parse a single option. */
static error_t parse_opt(int key, char *arg, struct argp_state *state)
{
    /* get the input argument from argp_parse, which we know is a
       pointer to our arguments structure. */
    struct arguments *arguments = state->input;

    switch (key)
    {
    case 'v':
        arguments->verbose = 1;
        break;
    case 'j':
        arguments->json = 1;
        break;
    case 'c':
        arguments->cmdline = 1;
        break;
    case 's':
        arguments->summary = 1;
        break;
    case ARGP_KEY_ARG:
        if (state->arg_num >= 2)
            /* too many arguments */
            argp_usage (state);
        arguments->args[state->arg_num] = atoi(arg);
        arguments->n_args++;
        break;
    default:
        return ARGP_ERR_UNKNOWN;
    }

    return 0;
}


void parse_args(int argc, char *argv[], struct arguments *arguments)
{
    /* program documentation */
    static char doc[] = "Report file descriptor statistics";
    /* positional arguments */
    static char args_doc[] = "INTERVAL COUNT";
    /* all options we accept */
    static struct argp_option options[] = {
        { "verbose", 'v', 0, 0, "Produce verbose output" },
        { "json", 'j', 0, 0, "Produce JSON output" },
        { "cmdline", 'c', 0, 0, "Output the command line instead of the PID" },
        { "summary", 's', 0, 0, "Display only totals (allocated and maximum)" },
        { 0 }
    };
    /* argp parser. */
    static struct argp argp = { options, parse_opt, args_doc, doc };

    /* parse arguments; every option seen by parse_opt will be
       reflected in the arguments struct */
    argp_parse(&argp, argc, argv, 0, 0, arguments);
}
